osimpipeline/task.py

```diff
@@ -96,29 +105,183 @@ def __init__(self, trial):
+        # Don't overwrite existing tasks file
+        if not os.path.exists(target[0]):
+            with open(file_dep[0]) as ft:
+                content = ft.read()
+                content = content.replace('@STUDYNAME@', self.study.name)
+                content = content.replace('@NAME@', self.trial.id)
+    
+            with open(target[0], 'w') as f:
+                f.write(content)
+
+    def copy_over_tasks(self):
+        if not os.path.exists(self.source_tasks_fpath):
+            # The user does not yet have a tasks.xml in place; fill out the
+            # template.
+            self.add_action(
+                    ['templates/%s/tasks.xml' % self.tool],
+                    [self.source_tasks_fpath],
+                    self.fill_tasks_template)
+            self.actions.append((self.copy_file,
+                [[self.source_tasks_fpath], [self.results_tasks_fpath]]))
+        else:
+            # We have already filled out the template tasks file,
+            # and the user might have made changes to it.
+            self.add_action(
+                    [self.source_tasks_fpath],
+                    [self.results_tasks_fpath],
+                    self.copy_file)
+
+    def generate_setup_file(self, cycle=None):
+        if cycle:
+            self.results_setup_fpath = os.path.join(self.tool_path, 
+                cycle.name, 'setup.xml')
+            self.add_action(
+                ['templates/%s/setup.xml' % self.tool],
+                [self.results_setup_fpath],
+                self.fill_setup_template,
+                init_time=cycle.start,
+                final_time=cycle.end
+                )
+        else:
+            self.results_setup_fpath = os.path.join(self.tool_path,'setup.xml')
+            self.add_action(
+                    ['templates/%s/setup.xml' % self.tool],
+                    [self.results_setup_fpath],
+                    self.fill_setup_template,  
+                    init_time=self.init_time,
+                    final_time=self.final_time,      
+                    )
+
 class ToolTask(TrialTask):
-    def __init__(self, trial, tool_folder, exec_name=None, cmd=None, env=None):
+    def __init__(self, tool_folder, trial, cycle=None,
+                 exec_name=None, cmd=None, env=None):
         super(ToolTask, self).__init__(trial)
-        self.name = '%s_%s' % (trial.id, tool_folder)
-        self.path = os.path.join(trial.results_exp_path, tool_folder)
+        self.tool_folder=tool_folder
+        self.exec_name=exec_name
+        self.cmd=cmd
+        self.env=env
+
+        if cycle:
+            self.name = '%s_%s_%s' % (trial.id, tool_folder, cycle.name)
+            self.path = os.path.join(trial.results_exp_path, tool_folder, cycle.name)
+            self.run_setup_file()
+        else:
+            self.name = '%s_%s' % (trial.id, tool_folder)
+            self.path = os.path.join(trial.results_exp_path, tool_folder)
+            self.run_setup_file()
```

Comment:
>This indicates to me that there needs to be a class in between TrialTask and ToolTask in the inheritance hierarchy.

Alternatively, as you suggested, we change `SetupTask` to support creation of individual cycle setup tasks, define `self.path` there so it can be reference through the setup task object, and avoid this problem altogether.